region/src/main/java/org/motechproject/nms/region/service/impl/LanguageLocationCodesImportServiceImpl.java

```diff
@@ -0,0 +1,219 @@
+            }
+        }
+    }
+
+    private void importRecord(Map<String, Object> record) {
+        String languageLocationCode = (String) record.get(LANGUAGE_LOCATION_CODE);
+        Language language = (Language) record.get(LANGUAGE);
+        State state = (State) record.get(STATE);
+        District district = (District) record.get(DISTRICT);
+        Circle circle = (Circle) record.get(CIRCLE);
+        Boolean defaultForCircle = (Boolean) record.get(DEFAULT_FOR_CIRCLE);
+
+        if (null != state && null != district) {
+            importRecordForStateAndDistrict(languageLocationCode, language, circle, defaultForCircle, state, district);
+        } else if (null != state) {
+            importRecordForState(languageLocationCode, language, circle, defaultForCircle, state);
+        } else if (null != district) {
+            importRecordForDistrict(languageLocationCode, language, circle, defaultForCircle, district);
+        } else {
+            throw new CsvImportDataException("Both state and district are null");
+        }
+    }
+
+    private void importRecordForStateAndDistrict(String languageLocationCode, Language language, Circle circle,
+                                                 Boolean defaultForCircle, State state, District district) {
+        verify(Objects.equals(district.getState().getCode(), state.getCode()),
+                "District's state does not match the state supplied in the CSV record");
+        verifyStateInCircle(circle, state);
+        verifyDistrictLanguageLocation(district);
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, Arrays.asList(district));
+    }
+
+    private void importRecordForState(String languageLocationCode, Language language, Circle circle,
+                                      Boolean defaultForCircle, State state) {
+        verifyStateInCircle(circle, state);
+        for (District stateDistrict : state.getDistricts()) {
+            verifyDistrictLanguageLocation(stateDistrict);
+        }
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, state.getDistricts());
+    }
+
+    private void importRecordForDistrict(String languageLocationCode, Language language, Circle circle,
+                                         Boolean defaultForCircle, District district) {
+        LOGGER.warn("State not specified for the '{}' district", district.getName());
+        verifyStateInCircle(circle, district.getState());
+        verifyDistrictLanguageLocation(district);
+        addDistrictToLanguageLocation(languageLocationCode, language, circle, defaultForCircle, Arrays.asList(district));
+    }
+
+    private void addDistrictToLanguageLocation(String languageLocationCode, Language language, Circle circle, Boolean defaultForCircle, Collection<District> districts) {
+        LanguageLocation languageLocation = languageLocationDataService.findByCode(languageLocationCode);
```

Comment:
Looking at the sample file languageLocationCode isn't actually unique.  I think if we change the domain object so circle <-> languageLocation becomes many to many than it can be unique again.   We'll have to make that change soon.